runtime/argument_validation.go

```diff
@@ -0,0 +1,354 @@
+	case interpreter.DictionaryDynamicType:
+		ok = valueConformsToDictionaryDynamicType(value, typ)
+	case interpreter.SomeDynamicType:
+		ok = valueConformsToSomeDynamicType(value, typ)
+
+	// Following types are guaranteed to be decoded correctly by the json.decode().
+	// However, this additional layer is added to protect the Cadence runtime,
+	// since the decoding currently relies on the host env, and it is external to Cadence.
+	case interpreter.StringDynamicType:
+		_, ok = value.(*interpreter.StringValue)
+	case interpreter.BoolDynamicType:
+		_, ok = value.(interpreter.BoolValue)
+	case interpreter.NumberDynamicType:
+		ok = valueConformsToSemaType(value, typ.StaticType)
+	case interpreter.NilDynamicType:
+		_, ok = value.(interpreter.NilValue)
+	case interpreter.AddressDynamicType:
+		_, ok = value.(interpreter.AddressValue)
+	case interpreter.MetaTypeDynamicType:
+		_, ok = value.(interpreter.TypeValue)
+	case interpreter.CapabilityDynamicType:
+		_, ok = value.(interpreter.CapabilityValue)
+	case interpreter.PrivatePathDynamicType, interpreter.PublicPathDynamicType, interpreter.StoragePathDynamicType:
+		_, ok = value.(interpreter.PathValue)
+
+	// Following types cannot be used as arguments to a script/transaction.
+	// However, still validate and allow wherever possible, so that this validation
+	// is less conservative and would seamlessly cater future changes.
+	case interpreter.FunctionDynamicType:
+		_, ok = value.(interpreter.FunctionValue)
+	case interpreter.BlockDynamicType:
+		_, ok = value.(interpreter.BlockValue)
+	case interpreter.DeployedContractDynamicType:
+		_, ok = value.(interpreter.DeployedContractValue)
+	case interpreter.StorageReferenceDynamicType:
+		// Currently this only checks whether the the value is `interpreter.StorageReferenceValue`.
+		// It doesn't check whether the 'referenced' value conforms to the innerType.
+		// TODO: add support for checking the referenced value conformance (if importing is supported).
+		_, ok = value.(*interpreter.StorageReferenceValue)
+	case interpreter.EphemeralReferenceDynamicType:
+		ok = valueConformsToEphemeralReferenceDynamicType(value, typ)
+	case interpreter.VoidDynamicType:
+		// Void type cannot have a value.
+		ok = false
+	}
+
+	return
+}
+
+func valueConformsToArrayDynamicType(value interpreter.Value, arrayType interpreter.ArrayDynamicType) bool {
+	arrayValue, ok := value.(*interpreter.ArrayValue)
+	if !ok || len(arrayValue.Values) != len(arrayType.ElementTypes) {
+		return false
+	}
+
+	for index, item := range arrayValue.Values {
+		if !valueConformsToDynamicType(item, arrayType.ElementTypes[index]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func valueConformsToDictionaryDynamicType(
+	value interpreter.Value,
+	dictionaryType interpreter.DictionaryDynamicType,
+) bool {
+
+	dictionaryValue, ok := value.(*interpreter.DictionaryValue)
+	if !ok || len(dictionaryValue.Keys.Values) != len(dictionaryType.EntryTypes) {
```

Comment:
Could be shortened, but OK as-is:

```suggestion
	if !ok || dictionaryValue.Count() != len(dictionaryType.EntryTypes) {
```